New attribute property keywords 1822

emanlove opened the pull request:
Added `Get DOM Attribute` and `Get Property` keywords and several acceptance tests. I know these test can be cleaned up a bit. One bit of review I am looking for is whether all the tests are there? Or if there might be something missing? I also added some comments within the tests. One issue I have been coming across is that when I go into an internal test (either acceptance or unit) some of them I have a tough time trying to see exactly what is is testing, and what should I expect it to do and verify. So I started to play around a bit with comments. I don't think there that great and need improvement. But this starts to highlight one idea for explaining tests better.

emanlove committed:
Initial commit adding keywords and tests for #1822

emanlove committed:
Added another atest to check for "attibute" that is both an attribute
and property

emanlove commented on atest/acceptance/keywords/elements.robot:
```diff
@@ -60,6 +60,55 @@ Get Element Attribute
+    [Setup]    Go To Page "forms/enabled_disabled_fields_form.html"
+    # Test get empty attribute
+    ${disabled}=    Get DOM Attribute    css:input[name="disabled_input"]    disabled
+    Should Be Equal    ${disabled}    true   # ${True}
```
The comment here at the end was what I was expecting the result to be; a boolean True. I left here for now to remind me that when I do the write up on how to use these new keywords to talk about what one should expect and to expect maybe the unexpected.

emanlove commented on atest/acceptance/keywords/elements.robot:
```diff
@@ -60,6 +60,55 @@ Get Element Attribute
+    ${disabled}=    Get DOM Attribute    css:input[name="disabled_input"]    disabled
+    Should Be Equal    ${disabled}    true   # ${True}
+    ${disabled}=    Get DOM Attribute    css:input[name="disabled_password"]    disabled
+    Should Be Equal    ${disabled}    true   # disabled
```
Same here; the comment here at the end was what I was expecting the result to be.

emanlove commented on atest/acceptance/keywords/elements.robot:
```diff
@@ -60,6 +60,55 @@ Get Element Attribute
+    ${empty_value}=    Get DOM Attribute    css:input[name="disabled_password"]    value
+    Should Be Equal    ${empty_value}    ${EMPTY}
+    ${disabled}=    Get DOM Attribute    css:input[name="enabled_password"]    disabled
+    Should Be Equal    ${disabled}    ${NONE}   # false
```
.. as with here.

emanlove commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -409,6 +409,37 @@ def get_element_attribute(
+        self, locator: Union[WebElement, str], attribute: str
+    ) -> str:
+        """Returns the value of ``attribute`` from the element ``locator``. `Get DOM Attribute` keyword
+        only returns attributes declared within the element's HTML markup.
```
The last sentence here,

> `Get DOM Attribute` keyword only returns attributes declared within the element's HTML markup.

come from the Selenium Python code. Technically this is not actually correct. I found with the tests above, and it would be good to have someone else verify, that if the requested attribute is not there it returns `${None}`. In part that was what I was exploring with the tests. Would it throw an error or return a NaN type response (which Python generally doesn't distinguish from None)? So I suspect we should change this to something else; still explaining what dom attribute is but clarify expected results.

yuriverweij replied:
Might be good to mention this behavior in the comments.

yuriverweij commented on atest/acceptance/keywords/elements.robot:
```diff
@@ -60,6 +60,55 @@ Get Element Attribute
+    # Test custom attribute
+    ${existing_custom_attr}=   Get DOM Attribute    id:emptyDiv  data-id
+    Should Be Equal    ${existing_custom_attr}    my_id
+    ${doesnotexist_custom_attr}=   Get DOM Attribute    id:emptyDiv  data-doesnotexist
```
I'd split these tests up (existing, existing custom attr, non existing custom attr) for the "normal" non-existing attr there is already a separate test below.

yuriverweij commented on atest/acceptance/keywords/elements.robot:
```diff
@@ -60,6 +60,55 @@ Get Element Attribute
+    ${children}=    Get Property    id:table1    children
+
+
+Get "Attribute" That Is Both An DOM Attribute and Property
```
Is it possible to add a test where the property is different than the attribute?

yuriverweij reviewed (commented):
Looks good to me. I've added a few comments.

emanlove committed:
Merge branch 'update-test-under-python-versions' into new-attribute-property-keywords-1822

emanlove committed:
Updated documentation for expected value if attribute is not there

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -27,7 +36,7 @@ jobs:
         pip install -r requirements-dev.txt
     - name: Install dependencies for pypy
-      if: matrix.python-version == 'pypy-3.7'
+      if: matrix.python-version == 'pypy-3.9'
```
Should something like the same comment as somewhere further on be added here? 
` # Recognize for the moment this will NOT run as we aren't using CPython 3.9`

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -36,25 +45,28 @@ jobs:
+        echo "$WEBDRIVERPATH"
     - name: Generate stub file for ${{ matrix.python-version }}
-      if: matrix.python-version != 'pypy-3.7'
+      if: matrix.python-version != 'pypy-3.9'
```
same here
` # Recognize for the moment this will NOT run as we aren't using CPython 3.9`

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -36,25 +45,28 @@ jobs:
 
     - name: Run tests with headless Chrome and with PyPy
-      if: matrix.python-version == 'pypy-3.7'
+      if: matrix.python-version == 'pypy-3.9'
```
and here 
` # Recognize for the moment this will NOT run as we aren't using CPython 3.9`

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -36,25 +45,28 @@ jobs:
 
-    - name: Run tests with normal Chrome and with Python 3.7
-      if: matrix.python-version == '3.7'
+    - name: Run tests with normal Chrome if CPython
```
This does not seem to be correct?  name says 'if CPython', but comparison in next line says '!= pypy-3-9'

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -65,12 +77,12 @@ jobs:
-        sudo chmod u+x ./selenium-server-standalone.jar
-        xvfb-run --auto-servernum python atest/run.py --zip headlesschrome  --grid True
+    # - name: Run tests with Selenium Grid
+    #   if: matrix.python-version == '3.11' && matrix.rf-version == '3.2.2' && matrix.python-version != 'pypy-3.7'
```
update to 'pypy-3-9' ?

yuriverweij commented on .github/workflows/CI.yml:
```diff
@@ -65,12 +77,12 @@ jobs:
-        wget --no-verbose --output-document=./selenium-server-standalone.jar http://selenium-release.storage.googleapis.com/3.141/selenium-server-standalone-3.141.59.jar
-        sudo chmod u+x ./selenium-server-standalone.jar
-        xvfb-run --auto-servernum python atest/run.py --zip headlesschrome  --grid True
+    # - name: Run tests with Selenium Grid
```
Should all this be commented out? These will not run anyway because 'matrix.rf-version 3.2.2' is not in the list. Maybe also add a comment like the others?

emanlove committed:
Cleaned up Get DOM Attribute, Get Property atests

emanlove committed:
Updated number of library keywords to 179

emanlove committed:
Revert "Updated number of library keywords to 179"

This reverts commit 4a1f413c4a4b29d196e76b8da5371019a23d3c32.

emanlove committed:
Updated number of library keywords to 179

emanlove committed:
Merge branch 'master' into new-attribute-property-keywords-1822

emanlove committed:
Added test where the property is different than the attribute

Using prefilled value on an input element, first verify that the value
attribute and property are the same. Then modifying the value by
inputting text verify the value property has changed but the attribute
value remains the same.

emanlove committed:
Merge branch 'master' into new-attribute-property-keywords-1822
